File: app/src/test/java/com/futsch1/medtimer/schedulertests/ReminderSchedulerActiveDayOfMonthUnitTest.java
package com.futsch1.medtimer.schedulertests;

import static com.futsch1.medtimer.schedulertests.ReminderSchedulerUnitTest.getScheduler;
import static com.futsch1.medtimer.schedulertests.TestHelper.on;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

import com.futsch1.medtimer.database.FullMedicine;
import com.futsch1.medtimer.database.Reminder;
import com.futsch1.medtimer.database.ReminderEvent;
import com.futsch1.medtimer.reminders.scheduling.ReminderScheduler;
import com.futsch1.medtimer.reminders.scheduling.ScheduledReminder;

import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

class ReminderSchedulerActiveDayOfMonthUnitTest {
    @Test
    void scheduleDayOfMonth() {
        ReminderScheduler scheduler = getScheduler(1);

        FullMedicine medicineWithReminders = TestHelper.buildFullMedicine(1, "Test");
        Reminder reminder = TestHelper.buildReminder(1, 1, "1", 480, 1);
        reminder.activeDaysOfMonth = 0x7;
        medicineWithReminders.reminders.add(reminder);

        List<FullMedicine> medicineList = new ArrayList<>();
        medicineList.add(medicineWithReminders);

        List<ReminderEvent> reminderEventList = new ArrayList<>();

        List<ScheduledReminder> scheduledReminders = scheduler.schedule(medicineList, reminderEventList);
        assertEquals(1, scheduledReminders.size());
        assertEquals(on(2, 480), scheduledReminders.get(0).timestamp());

        when(scheduler.getTimeAccess().localDate()).thenReturn(LocalDate.EPOCH.plusDays(2));
        scheduledReminders = scheduler.schedule(medicineList, reminderEventList);
        assertEquals(1, scheduledReminders.size());
        assertEquals(on(3, 480), scheduledReminders.get(0).timestamp());

        when(scheduler.getTimeAccess().localDate()).thenReturn(LocalDate.EPOCH.plusDays(10));
        scheduledReminders = scheduler.schedule(medicineList, reminderEventList);
        assertEquals(1, scheduledReminders.size());
        assertEquals(on(32, 480), scheduledReminders.get(0).timestamp());
    }
}

File: app/src/test/java/com/futsch1/medtimer/schedulertests/ReminderSchedulerActiveDayOfMonthUnitTest.kt
package com.futsch1.medtimer.schedulertests

import com.futsch1.medtimer.database.FullMedicine
import com.futsch1.medtimer.database.ReminderEvent
import com.futsch1.medtimer.reminders.TimeAccess
import com.futsch1.medtimer.schedulertests.ReminderSchedulerUnitTest.Companion.getScheduler
import com.futsch1.medtimer.schedulertests.TestHelper.buildFullMedicine
import com.futsch1.medtimer.schedulertests.TestHelper.buildReminder
import com.futsch1.medtimer.schedulertests.TestHelper.on
import org.junit.Test
import org.mockito.Mockito
import java.time.LocalDate
import java.time.ZoneId
import kotlin.test.assertEquals

internal class ReminderSchedulerActiveDayOfMonthUnitTest {
    @Test
    fun scheduleDayOfMonth() {
        val mockTimeAccess: TimeAccess = Mockito.mock<TimeAccess>()
        Mockito.`when`(mockTimeAccess.systemZone()).thenReturn(ZoneId.of("Z"))
        Mockito.`when`(mockTimeAccess.localDate()).thenReturn(LocalDate.EPOCH.plusDays(1))
        val scheduler = getScheduler(mockTimeAccess)

        val medicineWithReminders = buildFullMedicine(1, "Test")
        val reminder = buildReminder(1, 1, "1", 480, 1)
        reminder.activeDaysOfMonth = 7
        medicineWithReminders.reminders.add(reminder)

        val medicineList = mutableListOf<FullMedicine>()
        medicineList.add(medicineWithReminders)

        val reminderEventList = emptyList<ReminderEvent>()

        var scheduledReminders =
            scheduler.schedule(medicineList, reminderEventList)
        assertEquals(1, scheduledReminders.size)
        assertEquals(on(2, 480), scheduledReminders[0].timestamp)

        Mockito.`when`(mockTimeAccess.localDate())
            .thenReturn(LocalDate.EPOCH.plusDays(2))
        scheduledReminders = scheduler.schedule(medicineList, reminderEventList)
        assertEquals(1, scheduledReminders.size)
        assertEquals(on(3, 480), scheduledReminders[0].timestamp)

        Mockito.`when`(mockTimeAccess.localDate())
            .thenReturn(LocalDate.EPOCH.plusDays(10))
        scheduledReminders = scheduler.schedule(medicineList, reminderEventList)
        assertEquals(1, scheduledReminders.size)
        assertEquals(on(32, 480), scheduledReminders[0].timestamp)
    }
}
